src/resource_manager/gateway.py

```diff
@@ -0,0 +1,238 @@
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CreateGatewayError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+class GatewayResourceManager(ResourceManager[GenericNamespacedResource]):
+    """Kubernetes Ingress resource controller."""
+
+    def __init__(self, namespace: str, labels: LabelSelector, client: Client) -> None:
+        """Initialize the GatewayResourceManager.
+
+        Args:
+            namespace: Kubernetes namespace.
+            labels: Label to be added to created resources.
+            client: Initialized lightkube client.
+        """
+        self._ns = namespace
+        self._client = client
+        self._labels = labels
+        self._gateway_class_generic_resource = create_global_resource(
+            CUSTOM_RESOURCE_GROUP_NAME, "v1", GATEWAY_CLASS_RESOURCE_NAME, GATEWAY_CLASS_PLURAL
+        )
+        self._gateway_generic_resource = create_namespaced_resource(
+            CUSTOM_RESOURCE_GROUP_NAME, "v1", GATEWAY_RESOURCE_NAME, GATEWAY_PLURAL
+        )
+
+    @property
+    def _name(self) -> str:
+        """Returns "gateway"."""
+        return "gateway"
+
+    @property
+    def _namespace(self) -> str:
+        """Returns the kubernetes namespace.
+
+        Returns:
+            The namespace.
+        """
+        return self._ns
+
+    @property
+    def _label_selector(self) -> str:
+        """Return the label selector for resources managed by this controller.
+
+        Return:
+            The label selector.
+        """
+        return ",".join(f"{k}={v}" for k, v in self._labels.items())
+
+    def _set_gateway_class(
+        self, configured_gateway_class: str, resource: GenericNamespacedResource
+    ) -> None:
+        """Set the configured gateway class, otherwise the cluster's default gateway class.
+
+        Args:
+            configured_gateway_class: The desired gateway class name.
+            resource: The Ingress resource object.
+
+        Raises:
+            CreateGatewayError: When there's no available gateway classes
+        """
+        gateway_classes = tuple(self._client.list(self._gateway_class_generic_resource))
+
+        if not gateway_classes:
+            logger.error("Cluster has no available gateway class.")
+            raise CreateGatewayError("No gateway class available.")
+
+        gateway_class_names = (
+            gateway_class.metadata.name
+            for gateway_class in gateway_classes
+            if gateway_class.metadata
+        )
+        if configured_gateway_class not in gateway_class_names:
+            logger.error(
+                "Configured gateway class %s not present on the cluster.", configured_gateway_class
+            )
+            raise CreateGatewayError(f"Gateway class {configured_gateway_class} not found.")
+
+        resource.spec["gatewayClassName"] = configured_gateway_class
```

Comment:
It's more a charm config validation logic, it should be done in the `resource_definition.py`